factory/src/lib.rs

```diff
@@ -0,0 +1,555 @@
+use anchor_lang::prelude::*;
+use controller::{self, Controller, MintCtx, BurnCtx};
+use members::{self, Members, Merchant};
+use anchor_lang::solana_program::clock::{self, Clock};
+use anchor_spl::token::{self, TokenAccount, Mint, Token};
+
+declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
+
+fn is_merchant(
+    merchant: Pubkey,
+    merchant_state: Account<Merchant>,
+    member_state: Pubkey,
+    members: Pubkey
+) -> bool {
+    let merchant_state_pubkey = Pubkey::find_program_address(&[
+        b"merchant".as_ref(),
+        member_state.key().as_ref(),
+        merchant.key().as_ref()
+    ], &members);
+
+    if merchant_state.key() == merchant_state_pubkey.0 {
+        return true
+    } else {
+        return false
+    }
+}
+
+#[program]
+pub mod factory {
+    use anchor_spl::token::Burn;
+
+    use super::*;
+
+    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
+        let factory = &mut ctx.accounts.factory_state;
+        factory.controller_program = ctx.accounts.controller_program.key();
+        factory.controller_state = ctx.accounts.controller_state.key();
+        factory.admin = ctx.accounts.controller_program.key();
+        factory.mint_request_count = 0;
+        factory.burn_request_count = 0;
+
+        Ok(())
+    }
+
+    pub fn set_custodian_deposit_address(ctx: Context<SetCustodianDepositAddress>, deposit_address: String) -> Result<()> {
+        let members_state = ctx.accounts.member_state.clone();
+
+        if !is_merchant(ctx.accounts.merchant.key(), ctx.accounts.merchant_state.clone(), ctx.accounts.members.key(), ctx.accounts.member_state.key().clone()) {
+            return Err(Errors::SenderNotAuthorized.into())
+        }
+
+        if members_state.custodian == ctx.accounts.signer.key() {
+            return Err(Errors::SenderNotAuthorized.into())
+        }
+
+        if deposit_address.len() == 0 {
+            return Err(Errors::InvalidDepositAddress.into())
+        }
+        
+        let custodian_deposit_address = &mut ctx.accounts.custodian_deposit_address;
+        custodian_deposit_address.address = deposit_address;
+        custodian_deposit_address.bump = *ctx.bumps.get("custodian_deposit_address").unwrap();
```

Comment:
my personal preference is to consistently return an error instead of unwrapping. When devs or users see program panic it might give an impression of a bug or something. but not a big deal